File: CPP/Program/HalfHumanDemon.cpp
#pragma once
#include "Manusia.cpp"
#include "Demon.cpp"

class HalfHumanDemon : public Manusia, public Demon {
private:
    string Transformasi; // Atribut untuk menyimpan bentuk transformasi
public:
    // Konstruktor HalfHumanDemon
    HalfHumanDemon(
        string nama, int tinggi, int umur, Status s,
        string role, string sihir, int mana, string transformasi)
        : Karakter(nama, tinggi, umur, s),
          Manusia(nama, tinggi, umur, s, role),
          Demon(nama, tinggi, umur, s, sihir, mana),
          Transformasi(transformasi) {}

    // Setter & Getter untuk Transformasi
    void setTransformasi(string t) { Transformasi = t; }
    string getTransformasi() { return Transformasi; }

    // Override tampilkanInfo untuk menampilkan info HalfHumanDemon
    void tampilkanInfo() override {
        cout << "[Half Human Demon]" << endl;
        Karakter::tampilkanInfo();
        cout << "Role: " << Role
             << ", Sihir: " << Sihir
             << ", Mana: " << Mana
             << ", Transformasi: " << Transformasi << endl;
        cout << "---------------------------" << endl;
    }
};

File: Python/Program/HalfHumanDemon.py
from Manusia import Manusia   # import class Manusia
from Demon import Demon       # import class Demon
from Karakter import Karakter # import class Karakter (base utama)

# Multiple inheritance: HalfHumanDemon mewarisi Manusia & Demon
class HalfHumanDemon(Manusia, Demon):
    def __init__(self, nama, tinggi, umur, status, role, sihir, mana, transformasi):
        # panggil constructor Manusia
        Manusia.__init__(self, nama, tinggi, umur, status, role)
        # panggil constructor Demon
        Demon.__init__(self, nama, tinggi, umur, status, sihir, mana)
        self._transformasi = transformasi  # atribut khusus HalfHumanDemon

    # Setter & Getter transformasi
    def set_transformasi(self, t): self._transformasi = t
    def get_transformasi(self): return self._transformasi

    def tampilkan_info(self):
        print("[Half Human Demon]")
        Karakter.tampilkan_info(self)  # tampilkan info dasar dari Karakter
        # tampilkan atribut dari gabungan Manusia + Demon + HalfHumanDemon
        print(f"Role: {self._role}, Sihir: {self._sihir}, Mana: {self._mana}, Transformasi: {self._transformasi}")
        print("---------------------------")
